urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+          }
+          thisDiff.text = bestEdit;
+          if (bestEquality2.length() != 0) {
+            nextDiff.text = bestEquality2;
+          } else {
+            pointer.remove(); // Delete nextDiff.
+            nextDiff = thisDiff;
+            thisDiff = prevDiff;
+          }
+        }
+      }
+      prevDiff = thisDiff;
+      thisDiff = nextDiff;
+      nextDiff = pointer.hasNext() ? pointer.next() : null;
+    }
+  }
+
+  /**
+   * Given two strings, compute a score representing whether the internal
+   * boundary falls on logical boundaries.
+   * Scores range from 6 (best) to 0 (worst).
+   * @param one First string.
+   * @param two Second string.
+   * @return The score.
+   */
+  private int diff_cleanupSemanticScore(String one, String two) {
+    if (one.length() == 0 || two.length() == 0) {
+      // Edges are the best.
+      return 6;
+    }
+
+    // Each port of this function behaves slightly differently due to
+    // subtle differences in each language's definition of things like
+    // 'whitespace'.  Since this function's purpose is largely cosmetic,
+    // the choice has been made to use each language's native features
+    // rather than force total conformity.
+    char char1 = one.charAt(one.length() - 1);
+    char char2 = two.charAt(0);
+    boolean nonAlphaNumeric1 = !Character.isLetterOrDigit(char1);
+    boolean nonAlphaNumeric2 = !Character.isLetterOrDigit(char2);
+    boolean whitespace1 = nonAlphaNumeric1 && Character.isWhitespace(char1);
+    boolean whitespace2 = nonAlphaNumeric2 && Character.isWhitespace(char2);
+    boolean lineBreak1 = whitespace1
+        && Character.getType(char1) == Character.CONTROL;
+    boolean lineBreak2 = whitespace2
+        && Character.getType(char2) == Character.CONTROL;
+    boolean blankLine1 = lineBreak1 && BLANKLINEEND.matcher(one).find();
+    boolean blankLine2 = lineBreak2 && BLANKLINESTART.matcher(two).find();
+
+    if (blankLine1 || blankLine2) {
+      // Five points for blank lines.
+      return 5;
+    } else if (lineBreak1 || lineBreak2) {
+      // Four points for line breaks.
+      return 4;
+    } else if (nonAlphaNumeric1 && !whitespace1 && whitespace2) {
+      // Three points for end of sentences.
+      return 3;
+    } else if (whitespace1 || whitespace2) {
+      // Two points for whitespace.
+      return 2;
+    } else if (nonAlphaNumeric1 || nonAlphaNumeric2) {
+      // One point for non-alphanumeric.
+      return 1;
+    }
+    return 0;
+  }
+
+  // Define some regex patterns for matching boundaries.
+  private Pattern BLANKLINEEND
+      = Pattern.compile("\\n\\r?\\n\\Z", Pattern.DOTALL);
+  private Pattern BLANKLINESTART
+      = Pattern.compile("\\A\\r?\\n\\r?\\n", Pattern.DOTALL);
+
+  /**
+   * Reduce the number of edits by eliminating operationally trivial equalities.
+   * @param diffs LinkedList of Diff objects.
+   */
+  public void diff_cleanupEfficiency(LinkedList<Diff> diffs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 25. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.